### ISSUE TITLE
common/step_download: broken with some Windows's symbolic links

### ISSUE THREAD

**Issue by grv87** [mitchellh/packer#4323]
Packer version: 0.12.1

Host platform: Windows 7 SP1

Builder: `virtualbox-ovf`

Under the following conditions:
* `source_path` points somewhere inside symbolic link or junction (doesn't matter whether it's absolute or relative path)
* destination of that symbolic link/junction has `\\?\` prefix

I got error:
```
source_path is invalid: GetFileAttributesEx \\?\R:: The parameter is incorrect.
```

Example and log: https://gist.github.com/grv87/773b2b0db08bde45e440fbfcf0680008

This problem didn't exist in Packer 0.12.0.

**Comment by mwhooker** [comment]
I'm fairly sure this was introduced in #3982

@artburkart any thoughts on how to handle symlinks?

**Comment by artburkart** [comment]
I'll take a look. I don't have access to a windows machine, so I guess the best thing I can do is create a test case that should fail, and figure out how to fix it from there.

**Comment by artburkart** [comment]
I know I won't be able to attempt a `build` a box using the virtualbox image offered [here](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/#downloads), but it should be enough to help me get the `validate` command—and hopefully `build`—sorted out.

Anyway, thank you @grv87 very much for the [SSCCE](http://sscce.org/). I foolishly attempted to repro on a Windows 10 box before I realized you said you were on a Windows 7 SP1 box. I couldn't repro on Windows 10. I'll try Windows 7 soon.

**Comment by artburkart** [comment]
Hmmm, I attempted to repro on a Windows 7 SP1 box with packer 12.1 this evening but had no luck.

**Comment by artburkart** [comment]
Tonight I decided to give AWS Workspaces a try and see if I could repro the bug on a Windows 2008 workspace. No dice. @grv87 - have you tried running your commands on a different Windows 7 SP1 machine? I've tried three different versions of the Windows OS and failed to reproduce this bug.

**Comment by grv87** [comment]
@artburkart, the problem are links with `\\?\` prefix. Links without such prefix works OK. Does your test cover it?

I'll try other machines and give you feedback in 1-3 days.

**Comment by artburkart** [comment]
As a non-Windows user, I've struggled to understand the significance of `\\?\`. I've found some [information on the topic](https://msdn.microsoft.com/en-us/library/aa365247(VS.85).aspx#maxpath).

I understand that `mklink /J link \\?\%cd%` creates a ["hard link pointing to a directory"](http://www.howtogeek.com/howto/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/) and the `\\?\` means the path must be full and not relative, so the `%cd%` is required because it gets the full path for the current directory, but why this wouldn't work is beyond me. And as far as I can tell, it _does_ work, which is why I figured writing a test to cover a basic symlink would be sufficient.

If it satisfied, I will write some code that shells out and makes the various types of symlinks when the tests are only being ran on a Windows machine.

**Comment by grv87** [comment]
According to error message, problem is the call of `GetFileAttributesEx` function.
[Documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364946(v=vs.85).aspx) says that `\\?\` prefix is supported by Unicode version (`GetFileAttributesExW`) only. It looks like ANSI version (`GetFileAttributesExA`) is called which doesn't support `\\?\` prefix.
This call is somewhere inside Packer or Go libraries, I just can't find where exactly. And that call wasn't there in 0.12.0.

Also, the problem could be that some libraries are built as non-Unicode. Or, it could be regression inside that libraries.

**Comment by artburkart** [comment]
Welp, I spent a few more hours taking a crack at it, but I have yet to reproduce the bug nor can I find any evidence of `GetFileAttributesExA` in the source code or dependencies. Because having nothing to show for my effort felt kind of crappy, I decided to add the [code I mentioned above to my PR](https://github.com/artburkart/packer/blob/a8a7daa6f2eae317a1ae6bcc0b15ff1b761cf769/builder/virtualbox/ovf/config_test.go#L133-L149). When the tests are executed on a Windows machine, a link with the special `J` and `\\?\` is made. If my code is correct - something that is admittedly hard to tell since I couldn't figure out how to build packer on a Windows box - then no bug seems to exist.

**Comment by grv87** [comment]
I successfully reproduced bug under Windows 8 SP1, with Packer 0.12.1 and 0.12.2 too.

To fail there should be `source_path` box: `UbuntuServer16.04.1-amd64-kernel-vbox\UbuntuServer16.04.1-amd64-kernel.ovf`. I didn't put it in GIST because it's big. Please, use whatever box you have on hand.

**Comment by grv87** [comment]
I think the problem is in this row: https://github.com/mitchellh/packer/blob/master/common/config.go#L93

**Comment by artburkart** [comment]
You know what, for some reason I assumed the problem occurred before it ever even reached the OVF, so I never attempted with an OVF. Thanks for the help! I'll take another stab at it next week, since I'll be in transit these next couple days.

**Comment by grv87** [comment]
The problem in `filepath.EvalSymLinks` was fixed in Go 1.8.
Any plans to use modern Go version?

**Comment by rickard-von-essen** [comment]
> Any plans to use modern Go version?

@grv87 Like one from the future? The last Packer release was built with go 1.7 which is the current release. 1.8 will be released soon but currently is only [rc3](https://go.googlesource.com/go/+/go1.8rc3)

**Comment by artburkart** [comment]
I was able to reproduce the bug in my Windows VM. The bug was introduced when I started using the `DownloadableURL` function (as @grv87 has pointed out) to accommodate remote file downloads without reinventing the wheel. While I have introduced a regression, this means the bug has also always existed for ISO builds. I believe I can fix the bug by pulling in some of the golang source code changes that were made in golang 1.8, but I'm unsure whether this is the desired path to resolution.

@rickard-von-essen - thoughts?

**Comment by grv87** [comment]
Personally I can wait for release of Go 1.8.

**Comment by mwhooker** [comment]
I'll make sure to use go 1.8 for the next release. Shall we close this then if it's fixed in 1.8?

**Comment by artburkart** [comment]
Personally, I'd rather it be left open until I've confirmed the upgrade fixes it. Once I figure out how to build packer in a Windows environment, I'll follow up. Last time I tried, I was not successful. However, I wasn't using Cygwin.

**Comment by mwhooker** [comment]
@artburkart I would just use `XC_OS=windows ./scripts/build.sh` and then copy the .exe over

**Comment by artburkart** [comment]
@mwhooker - Thanks for the tip. I've verified the bug goes away when built with golang version 1.8.

**Comment by mwhooker** [comment]
closed via #4568
